### PR TITLE
Bring project back to life

### PR DESCRIPTION
**Pull request by Shinigami92**

### CODE REVIEW

**Comment by Shinigami92** [comment]
@zimme Ready to review and merge-ready

I suggest to read the commit messages and maybe check commit by commit to understand what I have done 🙂

Tests still failing, but didn't touched them so much

If you have any questions, feel free to ask

**Comment by zimme** [comment]
I'll try my best to review this tomorrow

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,5 @@
+*.log
 coverage/
 dist/
 node_modules/
 package-lock.json
-yarn.lock
```
Comment (zimme) [inline .gitignore:5]: The reason I added both `package-lock.json` and `yarn.lock` is that they are not respected by users of the package. It's only useful for applications. npm won't publish the lock file and yarn will ignore any lock files in packages.

Comment (Shinigami92) [reply .gitignore:5]: lock files are used to enable reproducible creation in ci

File: package.json [hunk 2]
```diff
@@ -4,39 +4,35 @@
     "url": "https://github.com/linterjs/core/issues"
   },
   "dependencies": {
-    "common-tags": "^1.8.0",
```
Comment (zimme) [inline package.json:7]: Using ~ will only allow PATCH version updates, while ^ will allow MINOR version updates. I guess I'm ok with only allowing PATCH updates, however if this package will be using `@linter/cli` in the future that will have to be using ^ as our packages will never have changes to PATCH as that will always be 0.

File: .gitignore (hunk 1 above)
Comment (zimme) [reply .gitignore:5]: Valid point, I'll allow it... :D

File: package.json (hunk 2 above)
Comment (Shinigami92) [reply package.json:7]: Oh sorry! you are right, but only for `dependencies`.
But we should use ~ for `devDependencies`

I'm always using ~ but I'm using them in frontend projects, and these are not dependencies for other projects ^^

File: package.json [hunk 3]
```diff
@@ -4,39 +4,35 @@
...
+    "jest-cli": "~26.1.0",
+    "jest-plugin-fs": "~2.9.0",
+    "jest-plugins": "~2.9.0",
+    "prettier": "2.0.5",
+    "rimraf": "~3.0.2",
+    "ts-jest": "~26.1.3",
+    "typescript": "~3.9.7"
   },
   "engines": {
-    "node": ">=8.0.0"
```
Comment (zimme) [inline package.json:39]: Any specific reason for 12.18? Does it have any features we depend on? Can't we just require `>=12.0.0`?

File: package.json [hunk 4]
```diff
@@ -55,10 +51,11 @@
   },
   "scripts": {
     "build": "tsc",
-    "check": "npm-check",
     "clean": "rimraf coverage dist",
     "clean:build": "rimraf dist",
     "clean:test": "rimraf coverage",
+    "format": "prettier --write .",
+    "lint": "echo Currently no linter installed",
```
Comment (zimme) [inline package.json:58]: I'm hoping these will be changed to `@linter/cli` in the future :D

File: src/format.test.ts [hunk 5]
```diff
@@ -14,8 +14,8 @@ describe("Format", () => {
   });
 
   describe("with @linter/provider-eslint", () => {
-    let format;
-    let linterAdapterPromiseList;
+    let format: FormatFunction;
+    let linterAdapterPromiseList: Array<Promise<any>>;
```
Comment (zimme) [inline src/format.test.ts:18]: This should be an array of promises that resolves to `LinterAdapter`

File: package.json (hunk 3 above)
Comment (Shinigami92) [reply package.json:39]: Nope, no reasons, just the latest from LTS
Will change it to 12.0 👍

File: package.json (hunk 4 above)
Comment (Shinigami92) [reply package.json:58]: Exactly 🥰

File: src/linter-provider.ts [hunk 6]
```diff
@@ -1,13 +1,13 @@
...
-import requireRelative = require("require-relative");
 import {
   LinterProviderLoadError,
   ModuleNotLinterProviderError,
   NoLinterProvidersError,
 } from "./errors";
 import { LinterFactory } from "./linter-adapter";
 import { logger } from "./logger";
+import readPkgUp = require("read-pkg-up");
+import requireRelative = require("require-relative");
```
Comment (zimme) [inline src/linter-provider.ts:10]: These where above the relative imports to group package imports before relative imports.

File: jest.config.js [hunk 7]
```diff
@@ -2,15 +2,15 @@ module.exports = {
   collectCoverage: true,
   coverageThreshold: {
     global: {
-      branches: 100,
-      functions: 100,
-      lines: 100,
-      statements: 100,
```
Comment (zimme) [inline jest.config.js:8]: If this change was to fix non passing tests, let's try and fix the tests later so we get 100 coverage as the only thing I actually want to test is this package's API and all permutations of using the API. I don't want to have tests of implementation details.

Comment (Shinigami92) [reply jest.config.js:8]: It was not my aim to fix test passing but 100% coverage is something that sometimes cannot be possible at all
A coverage of 85%-95% is totally valid

Lines like https://github.com/linterjs/core/blob/master/src/errors.ts#L4-L7 wont be covered (currently) 🤔

**Pull request merged**